membersrvc/ca/aca.go

```diff
@@ -0,0 +1,564 @@
+func (attrPair *AttributePair) GetValidFrom() time.Time {
+	return attrPair.validFrom
+}
+
+//SetValidFrom returns time which is valid from the pair.
+func (attrPair *AttributePair) SetValidFrom(date time.Time) {
+	attrPair.validFrom = date
+}
+
+//GetValidTo returns time which is valid to the pair.
+func (attrPair *AttributePair) GetValidTo() time.Time {
+	return attrPair.validTo
+}
+
+//SetValidTo returns time which is valid to the pair.
+func (attrPair *AttributePair) SetValidTo(date time.Time) {
+	attrPair.validTo = date
+}
+
+//ToACAAttribute converts the receiver to the protobuf format.
+func (attrPair *AttributePair) ToACAAttribute() *pb.ACAAttribute {
+	var from, to *google_protobuf.Timestamp
+	if attrPair.validFrom.IsZero() {
+		from = nil
+	} else {
+		from = &google_protobuf.Timestamp{Seconds: attrPair.validFrom.Unix(), Nanos: int32(attrPair.validFrom.UnixNano())}
+	}
+	if attrPair.validTo.IsZero() {
+		to = nil
+	} else {
+		to = &google_protobuf.Timestamp{Seconds: attrPair.validTo.Unix(), Nanos: int32(attrPair.validTo.UnixNano())}
+
+	}
+	return &pb.ACAAttribute{attrPair.attributeName, attrPair.attributeValue, from, to}
+}
+
+// NewACA sets up a new ACA.
+func NewACA() *ACA {
+	aca := &ACA{NewCA("aca", initializeACATables)}
+
+	return aca
+}
+
+func (aca *ACA) getECACertificate() (*x509.Certificate, error) {
+	raw, err := aca.readCACertificate("eca")
+	if err != nil {
+		return nil, err
+	}
+	return x509.ParseCertificate(raw)
+}
+
+func (aca *ACA) getTCACertificate() (*x509.Certificate, error) {
+	raw, err := aca.readCACertificate("tca")
+	if err != nil {
+		return nil, err
+	}
+	return x509.ParseCertificate(raw)
+}
+
+func (aca *ACA) fetchAttributes(id, affiliation string) ([]*AttributePair, error) {
+	// TODO this attributes should be readed from the outside world in place of configuration file.
+	var attributes = make([]*AttributePair, 0)
+	attrs := viper.GetStringMapString("aca.attributes")
+
+	var attrOwner *AttributeOwner
+
+	for _, flds := range attrs {
+		vals := strings.Fields(flds)
+		if len(vals) >= 1 {
+			val := ""
+			for _, eachVal := range vals {
+				val = val + " " + eachVal
+			}
+			attributeVals := strings.Split(val, ";")
+			if len(attributeVals) >= 6 {
+				attrPair, err := NewAttributePair(attributeVals, attrOwner)
+				if err != nil {
+					return nil, errors.New("Invalid attribute entry " + val + " " + err.Error())
+				}
+				if attrPair.GetID() != id || attrPair.GetAffiliation() != affiliation {
+					continue
+				}
+				if attrOwner == nil {
+					attrOwner = attrPair.GetOwner()
+				}
+				attributes = append(attributes, attrPair)
+			} else {
+				Error.Printf("Invalid attribute entry '%v'", vals[0])
+			}
+		}
+	}
+	return attributes, nil
+}
+
+func (aca *ACA) populateAttributes(attrs []*AttributePair) error {
+	tx, dberr := aca.db.Begin()
+	if dberr != nil {
+		return dberr
+	}
+	for _, attr := range attrs {
+		if err := aca.populateAttribute(attr); err != nil {
```

Comment:
Shall we pass to populateAttribute the db transaction? Otherwise the sql commands issued there will not be under the umbrella of tx.